File: server/chess/s.go
package chess

import "math"

type S Piece

var possiblePos = map[Side][][2]int{
	Top:    {{3, 0}, {5, 0}, {4, 1}, {3, 2}, {5, 2}},
	Bottom: {{3, 9}, {5, 9}, {4, 8}, {3, 7}, {5, 7}},
}

func (s *S) isPossiblePos(dest [2]int) bool {
	positions := possiblePos[s.Side]
	for _, pos := range positions {
		if pos[0] == dest[0] && pos[1] == dest[1] {
			return true
		}
	}
	return false
}

func (s *S) GetNextPositions(board *Board) [][2]int {
	currentX := s.Pos[0]
	currentY := s.Pos[1]

	var moves = [][2]int{{-1, -1}, {1, 1}, {-1, 1}, {1, -1}}
	var positions [][2]int
	for _, move := range moves {
		pos := [2]int{currentX + move[0], currentY + move[1]}
		if s.CanMove(pos, board) {
			positions = append(positions, pos)
		}
	}
	return positions
}

func (s *S) CanMove(dest [2]int, board *Board) bool {
	destX := dest[0]
	destY := dest[1]
	origX := s.Pos[0]
	origY := s.Pos[1]

	if s.isPossiblePos(dest) && math.Abs(float64(origX-destX)) == 1 && math.Abs(float64(origY-destY)) == 1 {
		return (*Piece)(s).CanPlaceAtDest(dest, board)
	}

	return false
}

func NewS(color Color, pos [2]int, side Side, key string) *Piece {
	piece := NewPiece("s", color, pos, side, key)
	s := (*S)(piece)
	piece.IPiece = s
	return piece
}

File: wasm-chinese-chess/src/piece/s.rs
use crate::{
    board::Board,
    shared::{in_board, in_nine_place, in_own_side, Color, Pos, Role, Side, HEIGHT, WIDTH},
};
use ndarray::iter::Windows;
use std::collections::HashMap;
use std::{cell, convert::TryInto};

use super::{IPiece, Piece, PieceFields};
use lazy_static::lazy_static;

#[derive(Default, Debug, Clone, PartialEq)]
pub struct S {
    piece: PieceFields,
}

impl IPiece for S {
    fn get_next_positions(&self, board: &Board) -> Vec<Pos> {
        let Pos(current_x, current_y) = *self.get_pos();
        let moves: [(isize, isize); 4] = [(-1, -1), (1, 1), (-1, 1), (1, -1)];
        let mut positions: Vec<Pos> = Vec::new();

        for (delta_x, delta_y) in moves.iter() {
            let pos = Pos(current_x + *delta_x, current_y + *delta_y);
            if self.can_move(&pos, board) {
                positions.push(pos)
            }
        }
        return positions;
    }

    fn can_move(&self, dest: &Pos, board: &Board) -> bool {
        let Pos(dest_x, dest_y) = *dest;
        let Pos(orig_x, orig_y) = *self.get_pos();

        if self.is_possible_pos(dest)
            && (orig_x - dest_x).abs() == 1
            && (orig_y - dest_y).abs() == 1
        {
            return self.piece.can_place_at_dest(dest, board);
        }

        return false;
    }

    fn get_pos(&self) -> &Pos {
        &self.piece.pos
    }

    fn get_color(&self) -> &crate::shared::Color {
        &self.piece.color
    }

    fn get_role(&self) -> &Role {
        &self.piece.role
    }

    fn set_pos(&mut self, pos: Pos) {
        self.piece.pos = pos;
    }
}

lazy_static! {
    static ref POSSIBLE_POS: HashMap<Side, [Pos; 5]> = {
        let mut m = HashMap::new();
        m.insert(
            Side::Top,
            [Pos(3, 0), Pos(5, 0), Pos(4, 1), Pos(3, 2), Pos(5, 2)],
        );
        m.insert(
            Side::Bottom,
            [Pos(3, 9), Pos(5, 9), Pos(4, 8), Pos(3, 7), Pos(5, 7)],
        );
        m
    };
}
impl S {
    pub fn new(color: Color, pos: Pos, side: Side /*key: String */) -> Self {
        S {
            piece: PieceFields::new(Role::RS, color, pos, side),
        }
    }

    fn is_possible_pos(&self, dest: &Pos) -> bool {
        let Pos(dest_x, dest_y) = dest;
        let positions = *(POSSIBLE_POS.get(&self.piece.side).unwrap()) as [Pos; 5];
        for Pos(x, y) in positions.iter() {
            if x == dest_x && y == dest_y {
                return true;
            }
        }
        false
    }
}

#[cfg(test)]
mod tests {

    use crate::{
        board::Board,
        piece::Piece,
        shared::{Color, Pos, Side},
        test_utils::{TestDataCanMove, TestDataGetNextPositions},
    };

    use super::*;

    #[test]
    fn test_get_next_positions() {
        let p1 = S::new(Color::Red, Pos(5, 9), Side::Bottom);
        let p2 = S::new(Color::Red, Pos(3, 5), Side::Bottom);
        let p3 = S::new(Color::Red, Pos(5, 5), Side::Bottom);
        // let

        let testcases = [TestDataGetNextPositions {
            pieces: [Piece::S(p1.clone())].to_vec(),
            piece: Piece::S(p1.clone()),
            next_positions: [Pos(4, 8)].to_vec(),
        }];

        for TestDataGetNextPositions {
            pieces,
            piece,
            next_positions,
        } in testcases
        {
            let board = Board::new(pieces);
            assert_eq!(board.get_next_positions(piece), next_positions.to_vec());
        }
    }

    #[test]
    fn test_can_move() {
        let p1 = S::new(Color::Red, Pos(5, 9), Side::Bottom);

        let testcases = [TestDataCanMove {
            pieces: [Piece::S(p1.clone())].to_vec(),
            piece: Piece::S(p1.clone()),
            pos: Pos(4, 8),
            expected: true,
        }];

        for TestDataCanMove {
            pieces,
            piece,
            pos,
            expected,
        } in testcases
        {
            let board = Board::new(pieces);
            assert_eq!(board.can_move(&piece, &pos), expected);
        }
    }
}
